File: src/drivers/ultrasonic_driver/scripts/ultrasonic_processor_node.py
#!/usr/bin/env python3
"""Process ultrasonic ranges into obstacle/min-range outputs."""

from __future__ import annotations

from typing import List

import rospy
from std_msgs.msg import Bool, Float32, Float32MultiArray


class UltrasonicProcessorNode:
    def __init__(self) -> None:
        self.safe_distance = float(rospy.get_param("~safe_distance_m", 0.2))
        self.ranges_topic = rospy.get_param("~ranges_topic", "/ultrasonic/ranges")

        self.min_pub = rospy.Publisher("/ultrasonic/min_range", Float32, queue_size=1)
        self.obs_pub = rospy.Publisher("/ultrasonic/obstacle", Bool, queue_size=1)

        rospy.Subscriber(self.ranges_topic, Float32MultiArray, self.ranges_cb, queue_size=1)
        rospy.loginfo("[ultrasonic] subscribe: %s", self.ranges_topic)

    def ranges_cb(self, msg: Float32MultiArray) -> None:
        if not msg.data:
            return
        values: List[float] = list(msg.data)
        if len(values) < 3:
            return

        front_values = [v for v in values[:3] if v > 0.0]
        if not front_values:
            min_range = 0.0
        else:
            min_range = min(front_values)

        has_obstacle = min_range > 0.0 and min_range < self.safe_distance

        self.min_pub.publish(Float32(data=min_range))
        self.obs_pub.publish(Bool(data=has_obstacle))


def main() -> None:
    rospy.init_node("ultrasonic_processor_node")
    UltrasonicProcessorNode()
    rospy.spin()


if __name__ == "__main__":
    main()

File: src/drivers/ultrasonic_driver/src/ultrasonic_processor_node.cpp
#include "ultrasonic_driver/ultrasonic_processor_node.hpp"
#include <algorithm>
#include <limits>

namespace ultrasonic_driver {

UltrasonicProcessorNode::UltrasonicProcessorNode()
    : Node("ultrasonic_processor_node") {
    // Declare and load parameters
    this->declare_parameter("safe_distance_m", 0.2);
    this->declare_parameter("ranges_topic", "/ultrasonic/ranges");

    safe_distance_ = this->get_parameter("safe_distance_m").as_double();
    ranges_topic_ = this->get_parameter("ranges_topic").as_string();

    // Setup subscribers and publishers
    ranges_sub_ = this->create_subscription<std_msgs::msg::Float32MultiArray>(
        ranges_topic_, 10,
        std::bind(&UltrasonicProcessorNode::rangesCallback, this, std::placeholders::_1));
    
    min_pub_ = this->create_publisher<std_msgs::msg::Float32>(
        "/ultrasonic/min_range", 10);
    
    obs_pub_ = this->create_publisher<std_msgs::msg::Bool>(
        "/ultrasonic/obstacle", 10);

    RCLCPP_INFO(this->get_logger(), "UltrasonicProcessorNode initialized");
    RCLCPP_INFO(this->get_logger(), "  safe_distance: %.2f m", safe_distance_);
    RCLCPP_INFO(this->get_logger(), "  ranges_topic: %s", ranges_topic_.c_str());
}

void UltrasonicProcessorNode::rangesCallback(const std_msgs::msg::Float32MultiArray::SharedPtr msg) {
    // Check if data is empty
    if (msg->data.empty()) {
        return;
    }

    // Need at least 3 values for front sensors (F, FL, FR)
    if (msg->data.size() < 3) {
        return;
    }

    // Extract front sensors only (indices 0, 1, 2)
    std::vector<float> front_values;
    for (size_t i = 0; i < 3 && i < msg->data.size(); ++i) {
        float v = msg->data[i];
        if (v > 0.0f) {
            front_values.push_back(v);
        }
    }

    // Calculate minimum range
    float min_range = 0.0f;
    if (!front_values.empty()) {
        min_range = *std::min_element(front_values.begin(), front_values.end());
    }

    // Detect obstacle: valid reading AND below safe distance
    bool has_obstacle = (min_range > 0.0f) && (min_range < static_cast<float>(safe_distance_));

    // Publish results
    std_msgs::msg::Float32 min_msg;
    min_msg.data = min_range;
    min_pub_->publish(min_msg);

    std_msgs::msg::Bool obs_msg;
    obs_msg.data = has_obstacle;
    obs_pub_->publish(obs_msg);
}

}  // namespace ultrasonic_driver

int main(int argc, char** argv) {
    rclcpp::init(argc, argv);

    try {
        auto node = std::make_shared<ultrasonic_driver::UltrasonicProcessorNode>();
        rclcpp::spin(node);
    } catch (const std::exception& e) {
        RCLCPP_ERROR(rclcpp::get_logger("ultrasonic_processor_node"),
                     "Exception in ultrasonic_processor_node: %s", e.what());
        rclcpp::shutdown();
        return 1;
    }

    rclcpp::shutdown();
    return 0;
}
